### ISSUE TITLE
ioctl() cmd argument should be unsigned long, not signed int

### ISSUE THREAD

**Issue by jpandre** [libfuse/libfuse#367]
On Linux, the ioctl() function is defined in <sys/ioctl.h> with a request argument defined as an **unsigned long** :
`extern int ioctl (int __fd, unsigned long int __request, ...) __THROW;`

The fuse kernel module passes it to the library as an **uint32_t** in the struct fuse_ioctl_in
`uint32_t	cmd;`

The library passes it to the file system code as a **signed int** :
(low level)
`void (*ioctl) (fuse_req_t req, fuse_ino_t ino, int cmd, void *arg,
		       struct fuse_file_info *fi, unsigned flags,
const void *in_buf, size_t in_bufsz, size_t out_bufsz);`
(high level)
`int (*ioctl) (const char *, int cmd, void *arg,
struct fuse_file_info *, unsigned int flags, void *data);`

Now, the values defined for the ioctl() command are meant to be unsigned, and, for instance FITRIM is defined as 0xc0185879 in <linux/fs.h>, which does not fit into a signed int.

As a consequence the clang compiler complains :

```
external/ntfs-3g/libntfs-3g/ioctl.c:398:7: error: overflow converting
case value to switch condition type (3222820985 to
18446744072637405305) [-Werror,-Wswitch]
        case FITRIM:
             ^
bionic/libc/kernel/uapi/linux/fs.h:169:16: note: expanded from macro 'FITRIM'
#define FITRIM _IOWR('X', 121, struct fstrim_range)
               ^
bionic/libc/kernel/uapi/asm-generic/ioctl.h:51:29: note: expanded from
macro '_IOWR'
#define _IOWR(type,nr,size) _IOC(_IOC_READ | _IOC_WRITE, (type), (nr),
(_IOC_TYPECHECK(size)))
                            ^
bionic/libc/kernel/uapi/asm-generic/ioctl.h:46:32: note: expanded from
macro '_IOC'
#define _IOC(dir,type,nr,size) (((dir) << _IOC_DIRSHIFT) | ((type) <<
_IOC_TYPESHIFT) | ((nr) << _IOC_NRSHIFT) | ((size) << _IOC_SIZESHIFT))
                               ^
1 error generated.
```

Defining cmd as unsigned in (*ioctl)() would add consistency in libfuse and would avoid forcing workarounds, even though defining cmd as an uint64_t in struct fuse_ioctl_in would lead to compatibility issues.

**Comment by Nikratio** [comment]
Thanks for the report! Unfortunately you are right, we can't change `struct fuse_ioctl_in` because it needs to stay compatible with past kernels. However, I think changing the signature of the `ioctl` callback is fine. Could you submit a pull request?

**Comment by trapexit** [comment]
`cmd` is a `uint32_t` so it should be `unsigned int`. If you did change it to unsigned long you have to do more than change the public function signature otherwise it'll sign extend any negative values.

**Comment by Nikratio** [comment]
I agree - we should change it to `unsigned int`. Are you referring to the bug title? That refers to the fact that userspace sends an `unsigned long` - so ideally it should be this type everywhere. But we can't change that.

**Comment by trapexit** [comment]
Yes. If you change the public signature in libfuse HL or LL alone the `do_ioctl` function still uses `int` so the compiler would sign extend if casting to `unsigned long`.

**Comment by Nikratio** [comment]
Fixed in 0c59e2b509
